pkg/kubelet/rkt/rkt.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2015 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rkt
+
+import (
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/credentialprovider"
+	"github.com/coreos/go-systemd/dbus"
+	"github.com/golang/glog"
+)
+
+const (
+	acversion            = "0.5.1"
+	kubernetesUnitPrefix = "k8s"
+	systemdServiceDir    = "/run/systemd/system"
+
+	rktDataDir             = "/var/lib/rkt"
+	rktLocalConfigDir      = "/etc/rkt"
+	rktMetadataServiceFile = "rkt-metadata.service"
+	rktMetadataSocketFile  = "rkt-metadata.socket"
+
+	unitKubernetesSection = "X-Kubernetes"
+	unitPodName           = "POD"
+	unitRktID             = "RktID"
+
+	dockerPrefix = "docker://"
+)
+
+const (
+	rktBinName = "rkt"
+
+	Embryo         = "embryo"
+	Preparing      = "preparing"
+	AbortedPrepare = "aborted prepare"
+	Prepared       = "prepared"
+	Running        = "running"
+	Deleting       = "deleting" // This covers pod.isExitedDeleting and pod.isDeleting.
+	Exited         = "exited"   // This covers pod.isExited and pod.isExitedGarbage.
+	Garbage        = "garbage"
+)
+
+const (
+	dockerAuthTemplate = `{"rktKind":"dockerAuth","rktVersion":"v1","registries":[%q],"credentials":{"user":%q,"password":%q}}`
+)
+
+var (
+	systemd *dbus.Conn
+	absPath string
+)
+
+// Runtime implements the ContainerRuntime for rkt. The implementation
+// uses systemd, so in order to run this runtime, systemd must be installed
+// on the machine.
+type Runtime struct {
+	systemd *dbus.Conn
+	absPath string
+	config  *Config
+	// TODO(yifan): Refactor this to be generic keyring.
+	dockerKeyring credentialprovider.DockerKeyring
+}
+
+func startMetadataService(systemd *dbus.Conn) error {
+	units, err := systemd.ListUnits()
+	if err != nil {
+		return err
+	}
+	for _, u := range units {
+		if u.Name == rktMetadataServiceFile {
```

Comment:
Another meta/future note: at some point, the rkt metadata service should probably become an external requirement like dockerd. It becomes an aspect of deploying a Kubernetes node, rather than something managed directly by the Kubelet.